From bb5580d51a8291f3514ef2351fb741c26d02b349 Mon Sep 17 00:00:00 2001
From: Joshua Cranmer <joshua.cranmer@intel.com>
Date: Fri, 22 Apr 2022 10:48:26 -0400
Subject: [PATCH 033/172] Remove more getPointerElementType queries.

Most of this tranche of calls were being used to disambiguate between different
built-ins that share the same or similar names.

The exception is the change in visitCallSPIRVPipeBuiltin, which is a convoluted
way of adding a cast to i8 addrspace(4)* if not already i8 addrspace(4)*, so it
is instead rewritten to check if a cast needs to be added without querying
getPointerElementType.

(cherry picked from commit f727f4d191a766c7c82b5f5ea90be052335fb01a)
---
 lib/SPIRV/OCLToSPIRV.cpp | 21 ++++++++++-----------
 lib/SPIRV/SPIRVToOCL.cpp | 26 +++++++++++++-------------
 2 files changed, 23 insertions(+), 24 deletions(-)

diff --git a/lib/SPIRV/OCLToSPIRV.cpp b/lib/SPIRV/OCLToSPIRV.cpp
index 45602510..728daaf8 100644
--- a/lib/SPIRV/OCLToSPIRV.cpp
+++ b/lib/SPIRV/OCLToSPIRV.cpp
@@ -1743,10 +1743,11 @@ static const char *getSubgroupAVCIntelOpKind(StringRef Name) {
       .StartsWith(kOCLSubgroupsAVCIntel::SICPrefix, "sic");
 }
 
-static const char *getSubgroupAVCIntelTyKind(Type *Ty) {
-  auto *STy = cast<StructType>(cast<PointerType>(Ty)->getPointerElementType());
-  auto TName = STy->getName();
-  return TName.endswith("_payload_t") ? "payload" : "result";
+static const char *getSubgroupAVCIntelTyKind(StringRef MangledName) {
+  // We're looking for the type name of the last parameter, which will be at the
+  // very end of the mangled name. Since we only care about the ending of the
+  // name, we don't need to be any more clever than this.
+  return MangledName.endswith("_payload_t") ? "payload" : "result";
 }
 
 static Type *getSubgroupAVCIntelMCEType(Module *M, std::string &TName) {
@@ -1778,11 +1779,9 @@ void OCLToSPIRVBase::visitSubgroupAVCBuiltinCall(CallInst *CI,
 
   // Update names for built-ins mapped on two or more SPIRV instructions
   if (FName.find(Prefix + "ime_get_streamout_major_shape_") == 0) {
-    auto PTy = cast<PointerType>(CI->getArgOperand(0)->getType());
-    auto *STy = cast<StructType>(PTy->getPointerElementType());
-    assert(STy->hasName() && "Invalid Subgroup AVC Intel built-in call");
-    FName += (STy->getName().contains("single")) ? "_single_reference"
-                                                 : "_dual_reference";
+    // _single_reference functions have 2 arguments, _dual_reference have 3
+    // arguments.
+    FName += (CI->arg_size() == 2) ? "_single_reference" : "_dual_reference";
   } else if (FName.find(Prefix + "sic_configure_ipe") == 0) {
     FName += (CI->arg_size() == 8) ? "_luma" : "_luma_chroma";
   }
@@ -1818,8 +1817,8 @@ void OCLToSPIRVBase::visitSubgroupAVCWrapperBuiltinCall(
   // Find 'to_mce' conversion function.
   // The operand required conversion is always the last one.
   const char *OpKind = getSubgroupAVCIntelOpKind(DemangledName);
-  const char *TyKind = getSubgroupAVCIntelTyKind(
-      CI->getArgOperand(CI->arg_size() - 1)->getType());
+  const char *TyKind =
+      getSubgroupAVCIntelTyKind(CI->getCalledFunction()->getName());
   std::string MCETName =
       std::string(kOCLSubgroupsAVCIntel::TypePrefix) + "mce_" + TyKind + "_t";
   auto *MCETy =
diff --git a/lib/SPIRV/SPIRVToOCL.cpp b/lib/SPIRV/SPIRVToOCL.cpp
index 3f137a59..25d5c389 100644
--- a/lib/SPIRV/SPIRVToOCL.cpp
+++ b/lib/SPIRV/SPIRVToOCL.cpp
@@ -580,10 +580,8 @@ void SPIRVToOCLBase::visitCallSPIRVPipeBuiltin(CallInst *CI, Op OC) {
         auto &P = Args[Args.size() - 3];
         auto T = P->getType();
         assert(isa<PointerType>(T));
-        auto ET = T->getPointerElementType();
-        if (!ET->isIntegerTy(8) ||
-            T->getPointerAddressSpace() != SPIRAS_Generic) {
-          auto NewTy = PointerType::getInt8PtrTy(*Ctx, SPIRAS_Generic);
+        auto *NewTy = PointerType::getInt8PtrTy(*Ctx, SPIRAS_Generic);
+        if (T != NewTy) {
           P = CastInst::CreatePointerBitCastOrAddrSpaceCast(P, NewTy, "", CI);
         }
         return DemangledName;
@@ -926,15 +924,17 @@ void SPIRVToOCLBase::visitCallSPIRVAvcINTELEvaluateBuiltIn(CallInst *CI,
         // reference image
         // 3. With single reference image - uses one OpVmeImageINTEL opcode for
         // reference image
-        int NumImages = std::count_if(Args.begin(), Args.end(), [](Value *Arg) {
-          if (auto *PT = dyn_cast<PointerType>(Arg->getType())) {
-            if (auto *ST = dyn_cast<StructType>(PT->getPointerElementType())) {
-              if (ST->getName().startswith("spirv.VmeImageINTEL"))
-                return true;
-            }
-          }
-          return false;
-        });
+        StringRef FnName = CI->getCalledFunction()->getName();
+        int NumImages = 0;
+        if (FnName.contains("SingleReference"))
+          NumImages = 2;
+        else if (FnName.contains("DualReference"))
+          NumImages = 3;
+        else if (FnName.contains("MultiReference"))
+          NumImages = 1;
+        else if (FnName.contains("EvaluateIpe"))
+          NumImages = 1;
+
         auto EraseVmeImageCall = [](CallInst *CI) {
           if (CI->hasOneUse()) {
             CI->replaceAllUsesWith(UndefValue::get(CI->getType()));
-- 
2.20.1

